### ISSUE TITLE
"Failed to load jshint library"

### ISSUE THREAD

**Issue by mikkotikkanen** [Microsoft/vscode-jshint#3]
Installation goes through successfully but after restarting I just get a error message:

`Failed to load jshint library. Please install jshint in your workspace folder using 'npm install jshint' and then press Retry.`

Runnning `npm install jshint` in extensions directory seems to fix this.

**Comment by replete** [comment]
i++;

**Comment by zachlysobey** [comment]
This is a bit confusing to me; I too get this message, and was thinking that it's because I don't have `jshint` installed (directly) at the *project* level. (In my project, I get `jshint` *transitively* through `gulp-jshint`.

Are you saying it's fixed if I add it in the extension directory? I tried that in `/Applications/Visual Studio Code.app/Contents/Resources/app/extensions`, but am still getting the same error.

**Comment by dbaeumer** [comment]
@zachlysobey the idea is that most devs that want to make use of jshint already have it installed. So we don't ship with a version. Instead we try to load it from the node_modules folder in the workspace folder (the folder open in VSCode) or the globally installed one. But I see the problem if jshint comes transitively through gulp-jshint. For now you can work around this by installing jshint globally.

**Comment by zachlysobey** [comment]
Thanks @dbaeumer, eep up the good work!

**Comment by riclf** [comment]
I'm not sure I understand why it is not following the VS Code extensions norm of loading from the home extension folder, at -
    Windows %USERPROFILE%\.vscode\extensions
    Mac $HOME/.vscode/extensions
    Linux $HOME/.vscode/extensions

Is that the ultimate intention ?  Thx.

**Comment by Jedarc** [comment]
+1

**Comment by zaus** [comment]
I installed jshint/eslint globally via the "Node.js command prompt" (Start > Node), which seems to have put it in  "%appdata%\npm\node_modules\eslint", but I still got this message; restarting a couple of times and it seems to have gone away?

**Comment by liuliangsir** [comment]
i encounter this problem too recently

**Comment by liuliangsir** [comment]
I encounter this problem too recently

**Comment by manish-tal** [comment]
I also encountered this issue
I have jshint installed globally via `npm install -g jshint`
but I always get this message.

**Comment by fasdgoc** [comment]
These lines below were needed to move from Default Settings to settings.json

    // JSHint configuration

    // Control whether or not jshint is enabled for JavaScript files.
    "jshint.enable": true,

    // A path to file containing the configuration options for jshint. If the file exists it overrides     jshint.options and any .jshintrc file
    "jshint.config": null,

    // The jshint options object to provide args to the jshint command.
    "jshint.options": {},

    // A path to a file containing patterns describing which files and directories should be ignored by  jshint. If the file exists it overrides jshint.exclude and any .jshintignore file.
    "jshint.excludePath": null,

    // Matching files and directories are being ignored by jshint. Glob patterns are interpreted relative to   the workspace's root folder.
    "jshint.exclude": {},

**Comment by finalclass** [comment]
I also have the same problem (running vscode on arch linux)

**Comment by Lichtjaeger** [comment]
You should follow your own advice (Microsoft/vscode-htmlhint#8):

> @mike-kaufman the rule should alwys be to prefer local installed libraries over bundled libraries. 
> So in your case if you bundle htmlhint the loading sequence should be:
> 
> -  use htmlhint from the workspace folder
> -  use global installed htmlhint
> -  use bundled htmlhint

Please bundle [a minified version](https://github.com/jshint/jshint/blob/master/dist/jshint.js) of JShint in your extension.

**Comment by leewinder** [comment]
"the idea is that most devs that want to make use of jshint already have it installed"
"Instead we try to load it from the node_modules folder in the workspace folder (the folder open in VSCode)"

Those two assumptions may not be true.  In my case, I do not (and will not) install modules globally and my node_modules folder is one folder deep, not at the root level of the project.

Either bundle a minified version of JShint in your extension as suggested by @Lichtjaeger or provide options to specify the install location.

**Comment by egamma** [comment]
@kieferrm version 2.5 of the https://github.com/Microsoft/vscode-languageserver-node provides additional configuration options for loading the library, see https://github.com/Microsoft/vscode-eslint/blob/master/eslint-server/src/server.ts#L209.

**Comment by jonathas** [comment]
+1 for providing options to specify the install location

**Comment by JesusIslam** [comment]
I too, have this problem even though I've already has jshint installed globally.

**Comment by wkaskie** [comment]
I just upgraded to release 1.17.1 and NOW I'm getting this error as well.

**Comment by Ashtheyogi** [comment]
I'm getting this error too. Can anyone write down the step by step process of fixing this? I dont understand how to npm install it or install it globally?

**Comment by scout208** [comment]
Just wanted to add you have to restart VS Code after running npm install if you already had it open.

**Comment by mike-kaufman** [comment]
> all cmd gives me is "'npm' is not recognized as an internal or external command, operable program or batch file." 

@21Vicroy - error above means npm insn't in your path.  Probably easiest thing to do is [install node](https://nodejs.org/en/download/), restart machine, and then try again.  Probably want to run `npm install -g jshint` (to install globally) or `npm  install jshint` in your project to install as part of a specific project.

**Comment by 21Vicroy** [comment]
@mike-kaufman how do I install it globally then? and how do I run it with windows 10? command prompt (admin) doesn't accept it.

**Comment by emanpatricio** [comment]
I'm usually using windows bash shell.

I fixed my issue by installing it in cmd and not in bash.

**Comment by melnakoury** [comment]
@emanpatricio Thanks man.. it's working for me now 👍

**Comment by di2pro** [comment]
I have a large project, that's split into micro services, the problem is that `node_modules` for a frontend part is nested in a few folders, because I use root folder. So, I don't wanna install anything globally (that's why I use docker all the time), I just wanna to specify the path for the `jshint` library in its config. Is it possible? Btw, it works fine in WebStorm for example, but WebStorm works a bit slower than VS Code because of indexing the code and so on.

**Comment by RMacfarlane** [comment]
@di2pro The latest version of the extension, 0.10.17, has a setting called `jshint.nodePath` that can be used for this. It can be an absolute path or relative path from the project root to your `node_modules` folder.

**Comment by RMacfarlane** [comment]
Closing this as there is now a `jshint.nodePath` setting for specifying the path to the `jshint` library. To see where `jshint` was loaded from, you can run the `JSHint: Show output` command when the extension is active, which will show the extension logs in the output panel. There's an info message there with the library path.

Please file a new issue if you're still having problems with this. Thanks!

**Comment by Imran4424** [comment]
how to install this globally?

**Comment by Lichtjaeger** [comment]
> 
> 
> how to install this globally?

`npm i -g jshint`

**Comment by lifenstein** [comment]
I am using nvm to change node versions on the fly when needed. The VS Code terminal picks it up just fine, but jshint doesn't seem to. I can hard-code the path for jshint to always use a specific version using `jshint.nodePath`, but I think it's cleaner if it recognized the path returned by `which node` or equivalent.

**Comment by dogbot** [comment]
>  `JSHint: Show output` command when the extension is active,

if the error is "Failed to load jshint library" how could JSHint be active? Also, "Run the command" where? in the terminal? On windows one could be on WSL or on the native CMD, which is it?  Utilities like JSHint are critical to people learning. These assumptions are maddening. Why not simply include JSHint? I have it installed both in WSL and in windows still it does not work

**Comment by Yuhemei** [comment]
1.windows+r
2. cmd
3.npm install -g jshint
4.Enter it
And then ,waiting for the installation.
Finally,,your problems will be solved

**Comment by Yuhemei** [comment]
1.windows+r
2. cmd
3.npm install -g jshint
4.Enter 
And then ,waiting for the installation.
Finally,,your problems will be solved

**Comment by Khobalt** [comment]
I'm using nvm and code doesn't pick up my global jshint, so running npm intall -g jshint didn't resolve this issue for me. So what I did, was head to the settings->Extensions->JSHint configuration and change this:
![image](https://user-images.githubusercontent.com/848683/66525395-33729d00-eaaa-11e9-981e-781fe9992aef.png)

I'm using a Mac, so to find out where my jshint lived I ran:
which jshint

**Comment by lantonov** [comment]
```
1.windows+r
2. cmd
3.npm install -g jshint
4.Enter
```
'npm' is not recognized as an internal or external command,
operable program or batch file.

**Comment by RMacfarlane** [comment]
@lantonov Do you have `npm` installed? (https://nodejs.org/en/download/) And is `C:\Program Files\nodejs\` part of your path? (https://stackoverflow.com/questions/33055515/npm-is-not-recognized-as-an-internal-or-external-command-operable-program-or)

**Comment by nektro** [comment]
Ran into this as well. Only loads from local install, not global.

**Comment by epmoyer** [comment]
I am having this issue and don't know how to debug it any further.
Running on OSX 10.14.6
- Completely uninstalled node and reinstalled using brew.
- From the command line: node works, npm works,  jshint  installs, and jshint can run/report its version:

```
bash-3.2$ which jshint
/usr/local/bin/jshint
bash-3.2$ jshint --version
jshint v2.11.0
```
- I've set the path explicitly in settings.json: `"jshint.nodePath": "/usr/local/bin/jshint",`
- I've uninstalled/reinstalled the jshint extension, cycled VSCode, and cycled the machine.

Still I get "Failed to load jshint library", and I don't know how to get any more information about why the extension *still* can't find it.

If I install jshint locally in the project then it works fine, but I need a global solution.

Thanks
